### ISSUE TITLE
LSTM Overfitting a small sample

### ISSUE THREAD

**Issue by michaelbzms** [slp-ntua/patrec-labs#73]
Για το LSTM πρεπει να μπορούμε να κάνουμε overfit σε ένα μικρό sample? Γιατί σε εμάς πχ για ένα batch με size 10 ξεκινάει με 2.3 loss, πέφτει κάπου στο 2.15 και μετά κολλάει στο ίδιο loss για τα υπόλοιπα epochs. Μπορεί αυτό να ευθύνεται στο ότι βρήκε κάποιο τοπικό ελάχιστο και παγιδεύτηκε εκεί? Είναι λογικό ή πρέπει να έχουμε κάποιο bug που το κάνει να μην μαθαίνει παραπάνω;

**Comment by filby89** [comment]
Γενικά πρέπει να μπορείτε να κάνετε overfit σε ένα μικρό sample. Π.χ. αν χρησιμοποιείτε μόνο 10 δείγματα και έχετε μεγάλο αριθμό παραμέτρων στα layer σας και δεν κάνει overfit πιθανώς να υπάρχει κάποιο σφάλμα στον κώδικα.

**Comment by michaelbzms** [comment]
Well, αν και νομίζω ότι είχα ξαναδοκιμάσει παρόμοιο configuration, με bidirectional, 2 layers και 64 rnn_size έκανε overfit. 

Όμως με όλο dataset δεν μαθαίνει και δεν ξέρω πόσο και πόσο άλλο να αυξήσω το capacity. Έχω φτάσει 6 layers, bidirectional 512 neurons rnn size και το loss πάει από 2.7 σε 2.2 και κολλάει εκεί (όλα τα accuracies κάπου στο 15-20%).

Έχετε κάποιο suggestion στα πόσα layers και size τους?

**Comment by michaelbzms** [comment]
Τελικά, δοκιμάζοντας πολλά και παρατηρόντας ότι για 1 batch μπορώ να κάνω overfit ενώ για παραπάνω όχι, το μόνο που βελτίωσε το training ήταν η χρήση του SequentialSampler αντι για του random στο dataset!

Πως ερμηνεύεται αυτό; Δηλαδή το ότι άλλαζαν τα batches σε κάθε epoch το εμπόδιζε να μάθει σε σχέση με το να είναι τα ίδια κάθε φορά;

**Comment by vassilispan** [comment]
> Τελικά, δοκιμάζοντας πολλά και παρατηρόντας ότι για 1 batch μπορώ να κάνω overfit ενώ για παραπάνω όχι, το μόνο που βελτίωσε το training ήταν η χρήση του SequentialSampler αντι για του random στο dataset!

> Πως ερμηνεύεται αυτό; Δηλαδή το ότι άλλαζαν τα batches σε κάθε epoch το εμπόδιζε να μάθει σε σχέση με το να είναι τα ίδια κάθε φορά;

Αυτό πρέπει να ισχύει γιατί όταν δεν κάνεις shuffle μεταξύ των εποχών, είναι ευκολότερο για το δίκτυο να μάθει κάποιο "μοτίβο" στα δεδομένα σου. Για αυτό γενικά είναι καλύτερο να κάνεις shuffle, καθώς το μοντέλο είναι κατά κάποιο τρόπο πιο ισχυρό.

Εμείς κάνουμε εκπαίδευση του μοντέλου με 2 layers, bidirectional και hidden size 500 (με χρήση RandomSampler) αλλά το evaluation κολλάει κοντά στο 13%. Υπάρχει κάποιο hint με το οποίο μπορούμε να προχωρήσουμε;

**Comment by michaelbzms** [comment]
> 
> 
> > Τελικά, δοκιμάζοντας πολλά και παρατηρόντας ότι για 1 batch μπορώ να κάνω overfit ενώ για παραπάνω όχι, το μόνο που βελτίωσε το training ήταν η χρήση του SequentialSampler αντι για του random στο dataset!
> 
> > Πως ερμηνεύεται αυτό; Δηλαδή το ότι άλλαζαν τα batches σε κάθε epoch το εμπόδιζε να μάθει σε σχέση με το να είναι τα ίδια κάθε φορά;
> 
> Αυτό πρέπει να ισχύει γιατί όταν δεν κάνεις shuffle μεταξύ των εποχών, είναι ευκολότερο για το δίκτυο να μάθει κάποιο "μοτίβο" στα δεδομένα σου. Για αυτό γενικά είναι καλύτερο να κάνεις shuffle, καθώς το μοντέλο είναι κατά κάποιο τρόπο πιο ισχυρό.

Καταλαβαίνω την ίδεα πίσω απο το randomness αλλά με αυτό κολλάει ενώ χωρίς αυτό μαθαίνει παραπάνω. Μάλιστα δεν πέφτει μόνο το train set αλλά και το validation loss (στο οποίο δεν γίνεται train). Παραδόξως, το test loss είναι λίγο χειρότερο (λίγο) αλλά το accuracy μεγαλύτερο.

Περιμένω απάντηση και από τους βοηθούς αλλά μέχρι στιγμής φαίνεται καλύτερο το μη random προκειμένου να μάθει παραπάνω έστω και με πιο biased τρόπο...

**Comment by SergeantChris** [comment]
#79 [comment](https://github.com/slp-ntua/patrec-labs/issues/79#issuecomment-764870069)
